### ISSUE TITLE
No threads when using MySQL?

### ISSUE THREAD

**Issue by Nonisiuniis** [alex/alchimia#36]
I'm writing app to consume event feed and store received events into mysql. Events should be routed to different tables based on a key in them and stored as soon as possible. But whatever i do, i can't get it to spawn more threads and performance is really bad for sequential inserts (~20 inserts/s).

multiple calls (params) => single thread, slow:
```
for item in data:
    d = engine.execute(table.insert().values(**item))
    d.addCallback(row_handler)
```

single call (multiparams) => single thread, fast:
```
d = engine.execute(table.insert(), data)
d.addCallback(all_handler)
```
I could possibly queue each event type and call multiparams style execute at regular intervals, but i'd rather use separate execute call for each event as soon as it arrives

I'm using version 0.7.0, mysql server 5.7, tested with pymysql, cymysql, mysqlconnector and mysqlclient, all the same.
When using adbapi, threads are created correctly

Am i missing something? Any ideas?

**Comment by glyph** [comment]
Hi @Nonisiuniis !

Sorry for the long delay.  Alchimia mostly just works and so its maintenance isn't very active :).

This behavior is at least somewhat intentional.  Each object in Alchimia (the TwistedEngine, and each TwistedConnection) is assigned a dedicated thread for all of its operations, to avoid sharing objects between different pool threads.

To get the behavior you want, call `.connect` to create a connection for every statement you want to execute in parallel.

Something like this (totally untested but hopefully you get the gist) should fix you up:

```python
def connected(connection, item):
    return connection.execute(table.insert().values(**item)).addCallback(row_handler).addBoth(lambda relay: connection.close().addCallback(lambda other: relay))
for item in data:
    engine.connect().addCallback(connected)
```

If you wanted to submit a patch that made `engine.execute` do this internally somehow, it would probably be accepted.  Personally I don't need this functionality so I'm not going to work on it, but feel free to open a new issue for it.
